### PR TITLE
feat: key value store backed by filesystem

### PR DESCRIPTION
**Pull request by yury-s**
Keys are used as a relative file path without any sanitization assuming that the underlying fs will throw on error.

### CODE REVIEW

File: docs/src/test-api/class-teststore.md [hunk 1]
```diff
@@ -0,0 +1,58 @@
+# class: TestStore
+* since: v1.31
+* langs: js
+
+Playwright Test provides a global `store` object for passing values between project test and tests. It is
```
Comment (dgozman) [inline docs/src/test-api/class-teststore.md:5]:
```suggestion
Playwright Test provides a global `store` object that can be used in global setup and tests.
```

File: docs/src/test-api/class-teststore.md [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+an error to call store methods outside of test and tests.
```
Comment (dgozman) [inline docs/src/test-api/class-teststore.md:6]:
```suggestion
```

File: docs/src/test-api/class-teststore.md [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+
+```js tab=js-js
+const { test, store } = require('@playwright/test');
+
+test('sign in', async ({ page, context }) => {
```
Comment (dgozman) [inline docs/src/test-api/class-teststore.md:11]: Let's do a beforeAll example with a lazy login instead?

File: docs/src/test-api/class-teststore.md [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+
+## async method: TestStore.get
+* since: v1.31
+- returns: <[any]>
+
+Get named item from the store. Returns undefined if there is no value with given name.
+
+### param: TestStore.get.name
+* since: v1.31
+- `name` <[string]>
```
Comment (dgozman) [inline docs/src/test-api/class-teststore.md:38]: Let's call it `path`?

File: docs/src/test-api/class-teststore.md [hunk 5]
```diff
@@ -0,0 +1,58 @@
+# class: TestStore
+* since: v1.31
+* langs: js
+
+Playwright Test provides a global `store` object for passing values between project test and tests. It is
+an error to call store methods outside of test and tests.
+
```
Comment (dgozman) [inline docs/src/test-api/class-teststore.md:7]: We'll need more docs :)

File: packages/playwright-test/src/index.ts [hunk 6]
```diff
@@ -22,9 +22,11 @@ import { createGuid, debugMode, removeFolders, addStackIgnoreFilter } from 'play
 import type { Fixtures, PlaywrightTestArgs, PlaywrightTestOptions, PlaywrightWorkerArgs, PlaywrightWorkerOptions, ScreenshotMode, TestInfo, TestType, TraceMode, VideoMode } from '../types/test';
 import type { TestInfoImpl } from './common/testInfo';
 import { rootTestType } from './common/testType';
+import { store as _baseStore } from './store';
 import { type ContextReuseMode } from './common/types';
 export { expect } from './matchers/expect';
 export const _baseTest: TestType<{}, {}> = rootTestType.test;
+export const store = _baseStore;
```
Comment (dgozman) [inline packages/playwright-test/src/index.ts:29]:
```suggestion
export { store } from './store';
```

File: packages/playwright-test/src/store.ts [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+import fs from 'fs';
+import path from 'path';
+import type { TestStore } from '../types/test';
+import { currentTestInfo } from './common/globals';
+
+class JsonStore implements TestStore {
+  private _toFilePath(name: string) {
+    const testInfo = currentTestInfo();
+    if (!testInfo)
+      throw new Error('store can only be called while test is running');
```
Comment (dgozman) [inline packages/playwright-test/src/store.ts:26]: We should also allow it during global setup. Perhaps even during describe, for test generation? Maybe no restrictions?

File: packages/playwright-test/src/store.ts [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+    const fileName = name + '.json';
+    return path.join(testInfo.config._internal.storeDir, fileName);
+  }
+
+  async get<T>(name: string) {
+    const file = this._toFilePath(name);
+    try {
+      const data = (await fs.promises.readFile(file)).toString('utf-8');
```
Comment (dgozman) [inline packages/playwright-test/src/store.ts:34]:
```suggestion
      const data = await fs.promises.readFile(file, 'utf-8');
```

File: tests/playwright-test/store.spec.ts [hunk 9]
```diff
@@ -127,44 +132,35 @@ test('should isolate store state between projects', async ({ runInlineTest }) =>
...
-        expect(await store.get('name')).toBe('str-' + test.info().project.name);
-      });
-    `,
-    'a.test.ts': `
+    'a.spec.ts': `
       const { test, store } = pwt;
-      test('should get data from setup', async ({ }) => {
-        expect(await store.get('number')).toBe(2022);
-        expect(await store.get('name')).toBe('str-' + test.info().project.name);
+      const storageKey = name => test.info().project.name + '/' + name;
```
Comment (dgozman) [inline tests/playwright-test/store.spec.ts:144]: I'd just drop this test, it doesn't really add value today.

**Review by dgozman — APPROVED** [review]

File: packages/playwright-test/src/store.ts (hunk 7 above)
Comment (yury-s) [reply packages/playwright-test/src/store.ts:26]: Let's address it separately.

File: tests/playwright-test/store.spec.ts (hunk 9 above)
Comment (yury-s) [reply tests/playwright-test/store.spec.ts:144]: yeah, I kept it to demonstrate how inconvenient it is at the moment.

**Pull request closed without merging**